### DISCUSSION TITLE
Deepgram doesn't seem to return the `request_id` immediately (pre-recorded)

### DISCUSSION THREAD

**Discussion by harshithpabbati** [deepgram/community#227] [General help] [answered]
### Which Deepgram product are you using?

Deepgram SDKs

### Details

I am using the Deepgram `preRecorded` method along with a callback URL. As per the documentation, it states that I should receive the `request_id` immediately before the audio processing begins.

However, there seems to be a significant delay in receiving the `request_id`, and it appears that it is actually provided only after the audio processing is completed.

### If you are making a request to the Deepgram API, what is the full Deepgram URL you are making a request to?

https://developers.deepgram.com/docs/callback

### If possible, please attach your code or paste it into the text box.

```
const response = await deepgram.transcription.preRecorded(
        { url: link },
        {
          model: 'whisper',
          punctuate: true,
          diarize: true,
          utterances: true,
          times: true,
          smart_format: true,
          callback: `myurl.com/webhook`,
        }
      );
```

**Comment by jkroll-deepgram** [comment]
Hi @harshithpabbati, your code worked for me. The `request_id` will be immediately assigned to the `response` variable, even if it takes more time to process the file and return a response to the callback URL. Here's a minimal working example in the Node interpreter:
```
> const { Deepgram } = require('@deepgram/sdk');
undefined
> const deepgram = new Deepgram('<API key>');
undefined
> await deepgram.transcription.preRecorded(
...         { url: 'https://static.deepgram.com/examples/fr_NatGen_CallCenter_BrokenWebsite.mp4' },
...         {
...           model: 'nova',
...           punctuate: true,
...           diarize: true,
...           utterances: true,
...           times: true,
...           smart_format: true,
...           callback: 'https://***.ngrok.io',
...         }
...       );
e { request_id: 'e0cd6187-3e91-4636-8e2a-1***********' }
```

Notice how the `request_id` gets logged right away (and would get immediately assigned to `const response` if specified), meanwhile later the callback URL will receive the response once the audio processing is complete.

**Reply by harshithpabbati** [reply]
Hi @jkroll-deepgram, it was fine with small files, but when I tried with an hour video, it took like 2 mins to just get the `request_id`

```
console.time("transcription");
  const response = await deepgram.transcription.preRecorded(
    {
      url: "url",
    },
    {
      model: "whisper",
      punctuate: true,
      diarize: true,
      utterances: true,
      smart_format: true,
      callback: `webhook_url`,
    }
  );
  console.log(response);
  console.timeEnd("transcription");
```

<img width="472" alt="Screenshot 2023-06-30 at 8 53 06 AM" src="https://github.com/deepgram/community/assets/43822585/5e52127d-bf0a-491e-a96f-e4a484352da2">

**Comment by jkroll-deepgram** [comment]
Hi @harshithpabbati, thanks for the follow-up info. I was able to replicate waiting several minutes for a request ID using your Node code. I believe the issue here is that to use the `deepgram.transcription.preRecorded()` method asynchronously, it needs to be wrapped in an async function. When I wrapped your code in `async function submitRequest() {}`, then I was able to go from several minutes down to 9 seconds to get a request ID response. Please give this a try and let me know if you see a similar speed improvement. I will also edit our Callback docs to make this clearer.
```
async function submitRequest() {
	console.time("transcription");
	const response = await deepgram.transcription.preRecorded(
	  {
	    url: "<audio-url>",
	  },
	  {
	    model: "whisper",
	    punctuate: true,
	    diarize: true,
	    utterances: true,
	    smart_format: true,
	    callback: "<callback-url>",
	  }
	);
	console.log(response);
	console.timeEnd("transcription");
}

submitRequest();
```

**Reply by harshithpabbati** [reply]
@jkroll-deepgram I was doing the same 🤔

**Reply by harshithpabbati** [reply]
I am seeing the same result even when I use the REST API (it's taking 1-2mins for larger files)

```
const transcribeMeeting = async () => {
  const headers = {
    accept: 'application/json',
    'content-type': 'application/json',
    Authorization: 'Token ' + process.env.DEEPGRAM_API_KEY,
  };
  const url = `https://api.deepgram.com/v1/listen?model=whisper&punctuate=true&diarize=true&smart_format=true&utterances=true&callback="mycallbackurl"`;
  const deepgramRes = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      url: "audio-url",
    }),
  });
  const res = await deepgramRes.json();
  console.log(res)
};
```

**Reply by jkroll-deepgram** [reply]
Can you try tweaking the first line? I see the minutes-long response with:
```
const transcribeMeeting = async () => {
```
and seconds-long response with:
```
async function transcribeMeeting() {
```

**Reply by harshithpabbati** [reply]
@jkroll-deepgram I have tried the same, but it's not really helping.

I could find the `request_id` in the deepgram console in seconds, but it's definitely taking a lot of time to return from the method/API.

**Comment by moskeyombus** [comment]
We seem to be experiencing this issue using a Ruby client to make API requests.  We had a request time out after 60 seconds due to failure to receive a response from Deepgram's API (which means we never received a `request_id` to store on our end and use to process the subsequent webhook).  The behavior does not appear to be consistent

We're extending our timeouts on our end to get around the issue, but this does incur cost so it's not ideal.  Here's an example of a request that completed successfully on your end, but took over 60 seconds to respond:
https://console.deepgram.com/project/a9a65da0-f047-4c3d-b710-c881b5444d10/usage/logs?page=1&start=2023-07-24&sel=628bf9e4-83ab-42ee-9b95-3275b75a1217

Here is our ruby code:
```
HTTParty.post('https://api.deepgram.com/v1/listen', { 
  headers: {
        'authorization' => "Token #{@api_key}",
        'content-type' => 'application/json'
      }, 
  query: {
        callback: '***',
        model: 'general',
        punctuate: true,
        smart_format: true,
        tier: 'nova',
        keywords: [...],
        diarize: true
  }, 
  body: { url: '***' }
})
```

**Reply by moskeyombus** [reply]
@jkroll-deepgram any idea here?  My ruby example clearly demonstrates that the issue on this ticket is not an issue with customer-written javascript, looking for some resolution.  We're experiencing this issue with our ruby clients, it's problematic in that, we never get a `request_id`, so we have no way to handle the webhook we eventually receive

**Reply by jkroll-deepgram** [reply]
@moskeyombus thanks for the mention to surface this. I got your code running, but am not yet replicating what you're reporting. I did make two small modifications to get the code running smoothly. Forgive me for being a total Ruby novice and perhaps these are obvious.

First, I initially got the error `Content-type was application/json, but we could not process the JSON payload.`, and needed to add `to_json` on to the body line to be: `body: { url: '***' }.to_json` in order for the JSON to be properly parsed.

Second, I added a print statement to print out the response, yielding the request ID. In the context of production code, I assume that would be assigned to a variable rather than just printed, but it gets the job done for this test snippet.

Here is the exact full code I'm running, with ngrok url and API key obscured:

```
require 'httparty'

print HTTParty.post('https://api.deepgram.com/v1/listen', { 
  headers: {
        'authorization' => 'Token <API-KEY>',
        'content-type' => 'application/json'
      }, 
  query: {
        callback: 'https://<...>.ngrok.io',
        model: 'general',
        punctuate: true,
        smart_format: true,
        tier: 'nova',
        diarize: true
  }, 
  body: { url: 'https://ia800305.us.archive.org/17/items/adventures_holmes/adventureholmes_01_doyle.mp3' }
})
```

This is using a 65-minute audio file, which is the longest I could quickly find freely available online for testing the URL JSON, to ensure that the audio would not be transcribed instantaneously, in order to observe the separation between returning the request ID and the response.

I then run `ruby test.rb` (having saved this code to a file), and within under a second, the program prints the request ID and completes. A few seconds later, I see the response JSON returned to the callback URL in another terminal window.

Does it sound like I'm doing exactly what you are, but getting different results? Typically I would inquire more about the details of your audio, but that sounds like it should be irrelevant in this case, given that the request ID should return immediately regardless of the audio or processing time. Please let me know if you spot anything I'm doing differently from you, so that I can continue attempting to reproduce.

**Reply by moskeyombus** [reply]
@jkroll-deepgram thanks for the feedback, yes you are doing exactly what we are doing.  Like I mentioned, most of the time we're getting the ID back relatively rapidly, it seems like occasionally there is a delay.  We've only experienced an outright failure to get a `request_id` once due to a 60 sec timeout.  

My intention here was to point out that the issue being described is not language-specific, it appears to be something behind your own API that occasionally will delay on response (containers spinning up? throttled file upload/download?).  Also worth mentioning that we typically deal with 4+ hour high-bitrate audio files, and the individual who opened this ticket mentioned using an hour video, both of which would be a much larger file than an hour audio file.  What's the file size on the file you are testing with?  Have you tested with anything over ~250mb?

**Comment by jkroll-deepgram** [comment, marked as answer]
@harshithpabbati @moskeyombus, sharing a belated update after our discussion on this topic a few months ago. As you were experiencing, we isolated a path in which async requests which provide an audio URL and a callback URL were lagging in returning a request ID due to attempting to download the audio synchronously to validate it, which can vary in speed due to fluctuating system load.

This issue has been resolved, so you should now observe that Deepgram first quickly checks that your audio is reachable, then promptly returns a request ID, then the full audio is downloaded, and any subsequent error message is returned to your callback URL if the audio fails to download after returning the request ID.

Please do let me know if you observe any re-occurrences of this issue, and share any request IDs, so that we can continue to improve our request handling and order of operations.
